src/todolistconverters.js

```diff
@@ -0,0 +1,321 @@
+		viewWriter.wrap( viewWriter.createRangeOn( checkbox ), label );
+
+		injectViewList( modelItem, viewItem, conversionApi, model );
+	};
+}
+
+/**
+ * A model-to-view converter for model `$text` element inside a todo list item.
+ *
+ * It is used by {@link module:engine/controller/datacontroller~DataController}.
+ *
+ * @see module:engine/conversion/downcastdispatcher~DowncastDispatcher#event:insert
+ * @param {module:utils/eventinfo~EventInfo} evt An object containing information about the fired event.
+ * @param {Object} data Additional information about the change.
+ * @param {module:engine/conversion/downcastdispatcher~DowncastConversionApi} conversionApi Conversion interface.
+ */
+export function dataModelViewTextInsertion( evt, data, conversionApi ) {
+	const parent = data.range.start.parent;
+
+	if ( parent.name != 'listItem' || parent.getAttribute( 'listType' ) != 'todo' ) {
+		return;
+	}
+
+	if ( !conversionApi.consumable.consume( data.item, 'insert' ) ) {
+		return;
+	}
+
+	const viewWriter = conversionApi.writer;
+	const viewPosition = conversionApi.mapper.toViewPosition( data.range.start );
+	const viewText = viewWriter.createText( data.item.data );
+	const span = viewWriter.createAttributeElement( 'span', { class: 'todo-list__label' } );
+	const label = viewWriter.createAttributeElement( 'label' );
+
+	viewWriter.insert( viewWriter.createPositionAt( viewPosition.parent, 'end' ), viewText );
+	viewWriter.wrap( viewWriter.createRangeOn( viewText ), span );
+	viewWriter.wrap( viewWriter.createRangeOn( viewText.parent ), label );
+}
+
+/**
+ * A view-to-model converter for checkbox element inside a view list item.
+ *
+ * It changes `listType` of model `listItem` to a `todo` value.
+ * When view checkbox is marked as checked the additional `todoListChecked="true"` attribute is added to model item.
+ *
+ * It is used by {@link module:engine/controller/datacontroller~DataController}.
+ *
+ * @see module:engine/conversion/upcastdispatcher~UpcastDispatcher#event:element
+ * @param {module:utils/eventinfo~EventInfo} evt An object containing information about the fired event.
+ * @param {Object} data An object containing conversion input and a placeholder for conversion output and possibly other values.
+ * @param {module:engine/conversion/upcastdispatcher~UpcastConversionApi} conversionApi Conversion interface to be used by the callback.
+ */
+export function dataViewModelCheckmarkInsertion( evt, data, conversionApi ) {
+	const modelCursor = data.modelCursor;
+	const modelItem = modelCursor.parent;
+	const viewItem = data.viewItem;
+
+	if ( viewItem.getAttribute( 'type' ) != 'checkbox' || modelItem.name != 'listItem' || !modelCursor.isAtStart ) {
+		return;
+	}
+
+	if ( !conversionApi.consumable.consume( viewItem, { name: true } ) ) {
+		return;
+	}
+
+	const writer = conversionApi.writer;
+
+	writer.setAttribute( 'listType', 'todo', modelItem );
+
+	if ( data.viewItem.getAttribute( 'checked' ) == 'checked' ) {
```

Comment:
This fails when upcasting checkboxes with binary attributes `input type="checkbox" checked` which are parsed as `checked = ""`. So this would need to check against an empty string as well.

Reply:
Wouldn't `data.viewItem.checked` be enough?

Reply:
It will be solved in the follow-up ticket https://github.com/ckeditor/ckeditor5-list/issues/138.

Reply:
@pomek Yes that should work

Great thanks! Also, awesome work on the tasklist :beers: